tests/test_metadata_eq_.py

```diff
@@ -0,0 +1,412 @@
+#!/usr/bin/env python
+
+# Copyright New York University and the TUF contributors
+# SPDX-License-Identifier: MIT OR Apache-2.0
+
+"""Test __eq__ implementations of classes inside tuf/api/metadata.py."""
+
+
+import copy
+import os
+import sys
+import unittest
+from typing import Any, ClassVar, Dict, List
+
+from securesystemslib.signer import SSlibSigner
+
+from tests import utils
+from tests.repository_simulator import RepositorySimulator
+from tuf.api.metadata import (
+    TOP_LEVEL_ROLE_NAMES,
+    DelegatedRole,
+    Delegations,
+    Key,
+    Metadata,
+    MetaFile,
+    Role,
+    TargetFile,
+)
+
+
+# Pylint no-member has to be disabled because of self.case_name.
+# pylint: disable=no-member
+class TestMetadataComparisions(unittest.TestCase):
+    """Test __eq__ for all classes inside tuf/api/metadata.py."""
+
+    metadata: ClassVar[Dict[str, bytes]]
+
+    @classmethod
+    def setUpClass(cls) -> None:
+        cls.repo_dir = os.path.join(
+            utils.TESTS_DIR, "repository_data", "repository", "metadata"
+        )
+        cls.metadata = {}
+        for md in TOP_LEVEL_ROLE_NAMES:
+            with open(os.path.join(cls.repo_dir, f"{md}.json"), "rb") as f:
+                cls.metadata[md] = f.read()
+
+    metadata_attributes_modifications: utils.DataSet = {
+        "signed": None,
+        "signatures": None,
+    }
+
+    @utils.run_sub_tests_with_dataset(metadata_attributes_modifications)
+    def test_metadata_eq_(self, test_case_data: Any) -> None:
+        md = Metadata.from_bytes(self.metadata["snapshot"])
+
+        # Assert that md is not equal to an object from another type
+        self.assertNotEqual(md, "")
+
+        md_2 = copy.deepcopy(md)
+        self.assertEqual(md, md_2)
+
+        setattr(md_2, self.case_name, test_case_data)
+        self.assertNotEqual(md, md_2)
```

Comment:
I agree that we are redoing steps 1-3 maany times and that it's probably to optimize them and I do agree with the second pseudocode and it should work.
There are two disadvantages to opting-out from the decorator: 
1. we don't receive an exact message which of the $DATA caused the effect as we do by using the decorator and a subtest
2. If one of the $DATA_SET cases fails then the other will do as well.

In a discussion with @lukpueh we remembered that the asserts provided `msg` option meaning that if an assert fails then the custom message will be appended. This will resolve **1** as a problem. For example, if we want to add `Failed case: {case}` to the message the result message of a failing equal looks like this: 
`<tuf.api.metadata.Metadata object at 0x7fdf6db05f00> != 'bbaba' : Failed case: signed`
meaning we still see log of the given argument.

So, the only downside to using a for loop here is 2 which I think is okay given that if for one of the attributes the test fails most likely it will fail for the others as well.

Reply:
We can still use the decorator. I think it makes the tests even cleaner (despite the repetitive first item in the data set rows). Here's an example how I aggregated the common parts of your `test_metadata_eq_`  and `test_signed_eq_`. You should be able to use these for most of your other test_*_eq functions as well. 

```python
   md_data: utils.DataSet = {
        "snapshot signed":       snapshot,           # test_metadata_eq_ (copy_and_simple_assert)
        "snapshot spec_version": snapshot["signed"], # test_signed_eq_ (copy_and_simple_assert)
        # add data for test_key_eq_, test_role_eq_, test_root_eq_, test_metafile_eq_, ...
        }

    @utils.run_sub_tests_with_dataset(md_data)
    def test_metadata__eq__(self, md: Any) -> None:
        self.assertNotEqual(md, "")
        md_2 = copy.deepcopy(md)
        self.assertEqual(obj, md)

    md_attrs_data: utils.DataSet = {
        "snapshot signed":       (snapshot, "signed", None),                     # test_metadata_eq_ (iteration 1)
        "snapshot signatures":   (snapshot, "signatures", None),                 # test_metadata_eq_ (iteration 2)
        "snapshot version":      (snapshot["signed"], "version", -1),            # test_signed_eq_ (iteration 1)
        "snapshot spec_version": (snapshot["signed"], "spec_version", "0.0.0"),  # test_signed_eq_ (iteration 2)
       # add data for test_key_eq_, test_role_eq_, test_root_eq_, test_metafile_eq_, ...
        }

    @utils.run_sub_tests_with_dataset(md_attrs_data)
    def test_metadata_attrs__eq__(self, test_case_data: Tuple) -> None:
        md, attr, val = test_case_data
        md_2 = copy.deepcopy(md)
        setattr(md_2, attr, val)
        self.assertNotEqual(md, md_2)
```

NOTE: I didn't include the initialization for the first argument (`md`), if you want to access what's in `cls.metadata`, you might need to define the test data sets in `setUpClass` as well.

Reply:
I honestly have problems figuring out if a specific test is even useful, eq() is such a special case. I don't have a strong opinion on the style

Reply:
Just wanted to say that your original idea of using the decorator was quite good, albeit with some restructuring. Apologies for causing any confusion. I'm really also fine either way. Thanks for your persistent efforts, @MVrachev!